File: scripts/bot.py
#!/usr/bin/env python3
"""
Telegram-бот Komorok для GitHub Actions.
Функции: /start, проверка сайта, отзывы, /admin.
"""

import os
import sys
import requests
import json
import hashlib
import subprocess

TOKEN = os.environ.get("BOT_TOKEN")
if not TOKEN:
    print("❌ BOT_TOKEN не найден")
    sys.exit(1)

SITE_URL = "https://komorok.ru"
TIMEOUT = 5
UPDATE_FILE = "last_update_id.txt"
FEEDBACK_FILE = "feedbacks.txt"
ADMIN_DOUBLE_HASH = os.environ.get("ADMIN_DOUBLE_HASH", "")


def git_commit_and_push(files):
    """Сохраняет изменения в GitHub."""
    try:
        result = subprocess.run(
            ["git", "status", "--porcelain"] + files,
            capture_output=True, text=True, timeout=10
        )
        if not result.stdout.strip():
            return
        subprocess.run(["git", "config", "user.name", os.environ.get("GIT_USER_NAME", "KomorokBot")], check=True)
        subprocess.run(["git", "config", "user.email", os.environ.get("GIT_USER_EMAIL", "bot@komorok.ru")], check=True)
        subprocess.run(["git", "add"] + files, check=True)
        subprocess.run(["git", "commit", "-m", "Обновлены данные бота"], check=True)
        subprocess.run(["git", "push"], check=True)
        print("✅ Изменения запушены")
    except Exception as e:
        print(f"❌ Git error: {e}")


def get_updates(offset):
    """Получает новые сообщения от Telegram."""
    url = f"https://api.telegram.org/bot{TOKEN}/getUpdates?offset={offset}&timeout=30"
    try:
        resp = requests.get(url, timeout=45)
        data = resp.json()
        return data.get("result", [])
    except:
        return []


def send_message(chat_id, text, reply_markup=None):
    """Отправляет сообщение в Telegram."""
    url = f"https://api.telegram.org/bot{TOKEN}/sendMessage"
    data = {"chat_id": chat_id, "text": text}
    if reply_markup:
        data["reply_markup"] = json.dumps(reply_markup)
    try:
        requests.post(url, data=data, timeout=10)
    except:
        pass


def check_site():
    """Проверяет, работает ли сайт."""
    try:
        resp = requests.get(SITE_URL, timeout=TIMEOUT, headers={"User-Agent": "KomorokBot/1.0"})
        return resp.status_code == 200
    except:
        return False


def handle_start(chat_id):
    """Обрабатывает команду /start."""
    keyboard = {
        "inline_keyboard": [
            [{"text": "🟢 Проверить сайт", "callback_data": "check"}],
            [{"text": "📝 Оставить отзыв", "callback_data": "feedback"}],
        ]
    }
    send_message(chat_id, "👋 Привет! Я бот Komorok. Выберите действие:", keyboard)


def handle_admin(chat_id, password=""):
    """Обрабатывает команду /admin."""
    if not password:
        send_message(chat_id, "🔒 Используйте: /admin <пароль>")
        return
    if not ADMIN_DOUBLE_HASH:
        send_message(chat_id, "❌ Хеш администратора не настроен")
        return

    first = hashlib.sha256(password.encode()).hexdigest()
    second = hashlib.sha256(first.encode()).hexdigest()

    if second != ADMIN_DOUBLE_HASH:
        send_message(chat_id, "❌ Неверный пароль!")
        return

    try:
        with open(FEEDBACK_FILE, "r", encoding="utf-8") as f:
            text = f.read()
        send_message(chat_id, f"📋 Отзывы:\n\n{text}" if text else "Нет отзывов")
    except:
        send_message(chat_id, "Нет отзывов")


def main():
    """Главный цикл бота."""
    # Читаем последний update_id
    try:
        with open(UPDATE_FILE, "r") as f:
            offset = int(f.read().strip()) + 1
    except:
        offset = 0

    updates = get_updates(offset)

    for update in updates:
        update_id = update["update_id"]
        offset = update_id

        # Обработка сообщений
        if "message" in update:
            msg = update["message"]
            chat_id = msg["chat"]["id"]
            text = msg.get("text", "").strip()

            if text == "/start":
                handle_start(chat_id)
            elif text.startswith("/admin"):
                password = text.removeprefix("/admin").strip()
                handle_admin(chat_id, password)
            elif text.startswith("/"):
                send_message(chat_id, "Неизвестная команда. Используйте /start")
            else:
                # Сохраняем как отзыв
                user = msg.get("from", {})
                name = user.get("first_name", "Аноним")
                username = user.get("username", "")
                with open(FEEDBACK_FILE, "a", encoding="utf-8") as f:
                    f.write(f"{name} (@{username}): {text}\n")
                send_message(chat_id, "💬 Спасибо за отзыв! Он сохранён.")
                git_commit_and_push([FEEDBACK_FILE])

        # Обработка нажатий на кнопки
        if "callback_query" in update:
            query = update["callback_query"]
            chat_id = query["message"]["chat"]["id"]
            data = query["data"]

            if data == "check":
                ok = check_site()
                send_message(chat_id, f"{'✅' if ok else '❌'} Сайт Komorok {'работает' if ok else 'не отвечает'}!")
            elif data == "feedback":
                send_message(chat_id, "📝 Напишите ваш отзыв о сайте (одним сообщением).")

    # Сохраняем offset
    with open(UPDATE_FILE, "w") as f:
        f.write(str(offset))


if __name__ == "__main__":
    main()

File: scripts/bot.kt
import java.io.*
import java.net.*
import java.security.MessageDigest
import org.json.JSONObject
import org.json.JSONArray

const val TOKEN = System.getenv("BOT_TOKEN") ?: error("BOT_TOKEN not set")
const val SITE_URL = "https://komorok.ru"
const val UPDATE_FILE = "last_update_id.txt"
const val FEEDBACK_FILE = "feedbacks.txt"

fun main() {
    val lastUpdateId = readLastUpdateId()
    val updates = getUpdates(lastUpdateId)
    
    for (update in updates) {
        val updateObj = update as JSONObject
        val updateId = updateObj.getInt("update_id")
        
        if (updateObj.has("message")) {
            val message = updateObj.getJSONObject("message")
            val chatId = message.getJSONObject("chat").getLong("id")
            val text = message.optString("text", "")
            
            when {
                text == "/start" -> handleStart(chatId)
                text.startsWith("/admin") -> handleAdmin(chatId, text.removePrefix("/admin").trim())
                text.startsWith("/") -> sendMessage(chatId, "Неизвестная команда. Используйте /start")
            }
        }
        
        if (updateObj.has("callback_query")) {
            val query = updateObj.getJSONObject("callback_query")
            val chatId = query.getJSONObject("message").getJSONObject("chat").getLong("id")
            val data = query.getString("data")
            
            when (data) {
                "check" -> handleCheck(chatId)
                "feedback" -> handleFeedbackPrompt(chatId)
            }
        }
        
        saveLastUpdateId(updateId)
    }
}

fun getUpdates(offset: Long): JSONArray {
    val url = URL("https://api.telegram.org/bot$TOKEN/getUpdates?offset=$offset&timeout=30")
    val conn = url.openConnection() as HttpURLConnection
    val response = conn.inputStream.bufferedReader().readText()
    val json = JSONObject(response)
    return json.optJSONArray("result") ?: JSONArray()
}

fun sendMessage(chatId: Long, text: String, replyMarkup: String? = null) {
    val url = URL("https://api.telegram.org/bot$TOKEN/sendMessage")
    val conn = url.openConnection() as HttpURLConnection
    conn.requestMethod = "POST"
    conn.setRequestProperty("Content-Type", "application/json")
    conn.doOutput = true
    
    val body = JSONObject()
    body.put("chat_id", chatId)
    body.put("text", text)
    if (replyMarkup != null) {
        body.put("reply_markup", JSONObject(replyMarkup))
    }
    
    conn.outputStream.write(body.toString().toByteArray())
    conn.inputStream.close()
}

fun handleStart(chatId: Long) {
    val keyboard = """
    {
        "inline_keyboard": [
            [{"text": "🟢 Проверить сайт", "callback_data": "check"}],
            [{"text": "📝 Оставить отзыв", "callback_data": "feedback"}]
        ]
    }
    """
    sendMessage(chatId, "👋 Привет! Я бот Komorok. Выберите действие:", keyboard)
}

fun handleCheck(chatId: Long) {
    val ok = try {
        val conn = URL(SITE_URL).openConnection() as HttpURLConnection
        conn.connectTimeout = 5000
        conn.readTimeout = 5000
        conn.responseCode == 200
    } catch (e: Exception) {
        false
    }
    sendMessage(chatId, if (ok) "✅ Сайт Komorok работает!" else "❌ Сайт не отвечает!")
}

fun handleFeedbackPrompt(chatId: Long) {
    sendMessage(chatId, "📝 Напишите ваш отзыв о сайте (одним сообщением).")
}

fun handleAdmin(chatId: Long, password: String) {
    if (password.isEmpty()) {
        sendMessage(chatId, "🔒 Используйте: /admin <пароль>")
        return
    }
    
    val hash = hashPassword(password)
    val expectedHash = System.getenv("ADMIN_DOUBLE_HASH") ?: ""
    
    if (hash != expectedHash) {
        sendMessage(chatId, "❌ Неверный пароль!")
        return
    }
    
    // Читаем отзывы
    val feedbacks = try {
        File(FEEDBACK_FILE).readText()
    } catch (e: Exception) {
        "Нет отзывов"
    }
    
    sendMessage(chatId, "📋 Отзывы:\n\n$feedbacks")
}

fun hashPassword(password: String): String {
    val first = MessageDigest.getInstance("SHA-256").digest(password.toByteArray())
        .joinToString("") { "%02x".format(it) }
    return MessageDigest.getInstance("SHA-256").digest(first.toByteArray())
        .joinToString("") { "%02x".format(it) }
}

fun readLastUpdateId(): Long {
    return try {
        File(UPDATE_FILE).readText().trim().toLongOrNull() ?: 0L
    } catch (e: Exception) {
        0L
    }
}

fun saveLastUpdateId(id: Long) {
    File(UPDATE_FILE).writeText(id.toString())
}